### PR TITLE
refactor: convert git datasources to class-based

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes:

Converts `git-refs` and `git-tags` datasources to be class-based. Also extracts `getRawRefs` into a common `GitDatasource` abstract class.

TODO:

  - [x] Figure out mocking in `lib/workers/repository/process/index.spec.ts` for `git-refs`

## Context:

#8647

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/git-refs/base.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    // fetch remote tags
+    const lsRemote = await git.listRemote([
+      getRemoteUrlWithToken(lookupName, hostType),
+    ]);
+    if (!lsRemote) {
+      return null;
+    }
+
+    const refMatch = /(?<hash>.*?)\s+refs\/(?<type>.*?)\/(?<value>.*)/;
+    const headMatch = /(?<hash>.*?)\s+HEAD/;
```
Comment (viceice) [inline lib/datasource/git-refs/base.ts:31]: Can we move these to module const for perf?

File: lib/datasource/git-tags/index.spec.ts [hunk 2]
```diff
@@ -51,7 +53,7 @@ describe('datasource/git-tags/index', () => {
           return Promise.resolve(lsRemote1);
         },
       });
-      const digest = await getDigest(
+      const digest = await new GitTagsDatasource().getDigest(
```
Comment (viceice) [inline lib/datasource/git-tags/index.spec.ts:56]: Do we really need a new instance on every test? or can we declare on module scope?

File: lib/datasource/git-refs/index.spec.ts [hunk 3]
```diff
@@ -59,7 +61,7 @@ describe('datasource/git-refs/index', () => {
           return Promise.resolve(lsRemote1);
         },
       });
-      const digest = await getDigest(
+      const digest = await new GitRefsDatasource().getDigest(
```
Comment (viceice) [inline lib/datasource/git-refs/index.spec.ts:64]: can we create a single instance instead? See also commont below.

File: lib/datasource/git-refs/index.ts [hunk 4]
```diff
@@ -1,110 +1,69 @@
...
-export async function getReleases({
-  lookupName,
-}: GetReleasesConfig): Promise<ReleaseResult | null> {
-  const rawRefs: RawRefs[] = await getRawRefs({ lookupName }, id);
+  @cache({
+    namespace: `datasource-${GitRefsDatasource.id}`,
+    key: ({ lookupName }: GetReleasesConfig) => lookupName,
+  })
+  // eslint-disable-next-line class-methods-use-this
+  async getReleases({
```
Comment (viceice) [inline lib/datasource/git-refs/index.ts:24]: can we add the override keyword? we can validate it later, when all other classes are prepared too.

File: lib/datasource/git-refs/base.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import simpleGit from 'simple-git';
+import { cache } from '../../util/cache/package/decorator';
+import { getRemoteUrlWithToken } from '../../util/git/url';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig } from '../types';
+import type { RawRefs } from './types';
+
+// TODO: extract to a separate directory structure (#10532)
+export abstract class GitDatasource extends Datasource {
```
Comment (viceice) [inline lib/datasource/git-refs/base.ts:9]: I don't see any benefit to extend from this class, as we only call one static method of it? 🤔

File: lib/datasource/git-tags/index.ts [hunk 6]
```diff
@@ -1,46 +1,60 @@
...
-  newValue?: string
-): Promise<string | null> {
-  const rawRefs = await gitRefs.getRawRefs({ lookupName }, id);
-  const findValue = newValue || 'HEAD';
-  const ref = rawRefs.find((rawRef) => rawRef.value === findValue);
-  if (ref) {
-    return ref.hash;
+  // eslint-disable-next-line class-methods-use-this
+  async getDigest(
+    { lookupName }: Partial<DigestConfig>,
```
Comment (viceice) [inline lib/datasource/git-tags/index.ts:49]: We shouldn't use `Partial` here

File: lib/datasource/git-refs/base.ts (hunk 5 above)
Comment (JamieMagee) [reply lib/datasource/git-refs/base.ts:9]: Right now, the main benefit is the shared common cache in the same cache namespace. Yes, I could do this with a single function, but I've written this now, and it allows us to share more common code between git refs and git tags easily in the future.

Comment (viceice) [reply lib/datasource/git-refs/base.ts:9]: I'm ok with having a shared class, bu i think we don't need to extend from it, as all static calls will be directly to base class, so there is nothing which can be overridden by extending.

So my idea is simlply use this class in a static manner, no datasource base class.

Then the other git datasources can call the static methods as usual. They don't need to extend this class to be able to call public statuic methods. It extending makes only sense for non-static member.

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by viceice — COMMENTED** [review]
Yes, lets do some real tests

**Review by rarkins — APPROVED** [review]

**Pull request merged**